synapse/handlers/auth.py

```diff
@@ -307,14 +309,26 @@ def _check_recaptcha(self, authdict, clientip):
                 defer.returnValue(True)
         raise LoginError(401, "", errcode=Codes.UNAUTHORIZED)
 
-    @defer.inlineCallbacks
     def _check_email_identity(self, authdict, _):
+        return self._check_threepid('email', authdict)
+
+    def _check_msisdn(self, authdict, _):
+        return self._check_threepid('msisdn', authdict)
+
+    @defer.inlineCallbacks
+    def _check_dummy_auth(self, authdict, _):
+        yield run_on_reactor()
+        defer.returnValue(True)
```

Comment:
Because these functions are all specified to return a deferred, so we need to yield somewhere in the function for it to actually return a deferred.